Overlay Synchronization

jbo023 opened the pull request:
next PR for the city of hamburg project see #494

we implemented an overviewsynchronizer

gberaudo commented on src/olcs/overlaysynchronizer.js:
```diff
@@ -0,0 +1,169 @@
+   * @type {ol.Collection.<ol.Overlay>}
+   * @private
+   */
+  this.overlays = this.map.getOverlays();
```
If it is private, it should have a trailing _.

gberaudo commented on src/olcs/overlaysynchronizer.js:
```diff
@@ -0,0 +1,169 @@
+  this.scene.canvas.parentElement.appendChild(this.overlayContainerStopEvent_);
+
+  /**
+   * @type {!Object<?,olcs.Overlay>}
```
Shouldn't it be `{!Object<string,olcs.Overlay>}`?

gberaudo commented on src/olcs/overlay.js:
```diff
@@ -0,0 +1,350 @@
+  this.scene = options.scene;
+
+  /** @type {!olcs.OverlaySynchronizer} */
+  this.synchronizer = options.synchronizer;
```
These 2 properties should be private.

gberaudo commented on src/olcs/overlay.js:
```diff
@@ -0,0 +1,350 @@
+   * @type {boolean}
+   * @todo read from parent
+   */
+  this.stopEvent_ = true;
```
What are you going to do about the todos?

gberaudo commented on src/olcs/overlay.js:
```diff
@@ -0,0 +1,350 @@
+   * @private
+   * @type {!ol.Overlay}
+   */
+  this.parent_ = options['parent'];
```
The quotes should be removed, see comment above.

gberaudo commented on src/olcs/overlay.js:
```diff
@@ -0,0 +1,350 @@
+    this.scenePostRenderListenerRemover_ = scene.postRender.addEventListener(this.updatePixelPosition.bind(this));
+    this.updatePixelPosition();
+    const container = this.stopEvent_ ?
+      this.synchronizer.getOverlayContainerStopEvent() : this.synchronizer.getOverlayContainer(); // TODO respect stop-event flag in synchronizer
```
What about this todo?

jbo023 commented on src/olcs/overlay.js:
```diff
@@ -0,0 +1,350 @@
+  const pixel = [pixelCartesian.x, pixelCartesian.y];
+  const mapSize = [scene.canvas.width, scene.canvas.height];
+  const that = /** @type {ol.Overlay} */ (this);
+  ol.Overlay.prototype.updateRenderedPosition.call(that, pixel, mapSize);
```
Yeah we also thought about this, but for this to work we have to either change the Overlay property visibility to "protected" from "private", or we can disable the closure compiler "visibility" Errors like this in the olcesium.json. This is also why we didn't do some of the todos in the code.

"jscomp_off": [
      "lintChecks",
      "analyzerChecks",
      "visibility"
    ],

So, what is your preference, should i disable the compiler flag, or should i create a openlayers PR which changes all ol.Overlay "@private" to "@protected". Then i can refactor the olcs.overlay to directly inherit from ol.Overlay.

gberaudo replied:
You should send a PR to OpenLayers to change the visibility to `@protected`. It is something we do when it appears there is a real use-case for it.

jbo023 replied:
Hmm, i just checked this, changing properties to @protected is unfortunately not enough because i need to access all the properties of the original Overlay in the new olcs.Overlay. To do this we have three options
- change propertie visibility ( insertFirst_, stopEvent_, ect.) to @public
- add getters properties which do not have them yet 
- add getOptions function which returns an options Object (my preferred  solution)

Which option do you prefer?

jbo023 commented:
Hi, 

i did some refactoring of the overlay synchronization. I renamed it to synchronizedOverlay and now inherit from ol/overlay.js. I think i also fixed all of your remarks, if still applicable. 
Jannes

jbo023 commented:
This only works after the referenced openlayers PR is merged. For tests the file /node_modules/openlayers/src/ol/overlay.js can be changed with the changes from the PR

gberaudo commented on Cesium.externs.js:
```diff
@@ -3416,3 +3444,37 @@ Cesium.EntityView.prototype.update = function(currentTime, bs) {};
+Cesium.CullingVolume = function(planes) {};
+
+/**
+ * @param {Object} boundingVolume
```
The type should be `Cesium.BoundingSphere`.

jbo023 replied:
I took the api description from the cesium API: https://cesiumjs.org/Cesium/Build/Documentation/CullingVolume.html
So computeVisibility works with every object which implements intersectsPlane().

gberaudo replied:
OK, then fine with it since it would be too cumbersome to introduce interfaces.

gberaudo commented:
Thanks, great. Waiting for the OpenLayers PR to be merged.

gberaudo commented:
@jbo023, could you please rebase and change package.json to depend on the updated OpenLayers version?
```
    "openlayers": "git+https://github.com/openlayers/openlayers#1a1d45fdd",
```

jbo023 commented:
@gberaudo 

rebase done, and i changed the openlayers dependency.

gberaudo commented:
I will now merge, thanks @jbo023 for your contribution.
